src/System.Data.SqlClient/src/System/Data/SqlClient/SqlParameter.cs

```diff
@@ -1978,6 +1963,32 @@ internal static string[] ParseTypeName(string typeName, bool isUdtTypeName)
             }
         }
 
+        /// <summary>
+        /// This method will decide whether the lazy allocated backing object for_xmlSchemaCollection is needed
+        /// and create it if it is needed and is currently null. The return value indicates whether the caller
+        /// should assign the value to the desired property
+        /// </summary>
+        /// <param name="value">the value which is going to be set to one of the properties</param>
+        /// <returns>true means that the assignment is needed. false means that it would result in no visible change</returns>
+        private bool ShouldAssignXmlSchemaCollectionValue(string value)
+        {
+            bool isNull = _xmlSchemaCollection is null;
+            if (!string.IsNullOrEmpty(value))
+            {
+                // not null or empty so we need to make a place to put it if we don't already have one
+                if (isNull)
+                {
+                    _xmlSchemaCollection = new SqlMetaDataXmlSchemaCollection();
+                }
+                return true;
+            }
+            else
+            {
+                // value is null or empty so we only need to assign if we might already have a value to overwrite
+                return !isNull;
```

Comment:
What happens if collection is not null, and value is null?

Reply:
if value is null and collection exists it will return true from the !isNull check which will make the caller assign. So if there is an object already always write to it.

Reply:
So you will set `null` value on database? Is that correct behavior?

Reply:
I'm asking to get some clarity on what values are expected to be assigned.

If `null` is allowed for `value`, why check `value` at the first place and if it is not, then whether or not `collection` is null, it shouldn't set database to `null`. This implementation is very contradictory right now.

Reply:
The getters turn null into empty strings so assigning null or empty string is only useful if there is currently a value. The code to make this lazy initialization just seems to be messy no matter how I try to express it so I understand your concern.

The intention behind the change is to prevent 3 string refs being allocated in SqlParameter when they're very rarely used by using the lazy container object. The desire is to create the container object only if it's needed (a non-null, non-empty value needs to be stored) and then if one exists use it normally since we've hit a code path that needs those three strings for something.

Reply:
I added a test which may help understand it. 
```csharp
SqlParameter parameter = new SqlParameter(); 

Assert.Equal(string.Empty, parameter.XmlSchemaCollectionName); 

// verify that if we set it to null we still get an empty string back 
parameter.XmlSchemaCollectionName = null; 
Assert.Equal(string.Empty, parameter.XmlSchemaCollectionName); 

// verify that if we set a value we get it back 
parameter.XmlSchemaCollectionName = "name"; 
Assert.Equal("name", parameter.XmlSchemaCollectionName); 

// verify that if we set it explicitly to null it reverts to empty string 
parameter.XmlSchemaCollectionName = null; 
Assert.Equal(string.Empty, parameter.XmlSchemaCollectionName);
```